src/pose-sensor.js

```diff
@@ -0,0 +1,180 @@
+const X_AXIS = new Vector3(1, 0, 0);
+const Z_AXIS = new Vector3(0, 0, 1);
+
+let orientation = {};
+if (screen.orientation) {
+  orientation = screen.orientation;
+} else if (screen.msOrientation) {
+  orientation = screen.msOrientation;
+} else {
+  Object.defineProperty(orientation, 'angle', {
+    get: () => { return (window.orientation || 0); }
+  });
+}
+
+// Quaternion to rotate from sensor coordinates to WebVR coordinates
+const SENSOR_TO_VR = new Quaternion();
+SENSOR_TO_VR.setFromAxisAngle(X_AXIS, -Math.PI / 2);
+SENSOR_TO_VR.multiply(new Quaternion().setFromAxisAngle(Z_AXIS, Math.PI / 2));
+
+/**
+ * An abstraction class around either using the new RelativeOrientationSensor,
+ * or `devicemotion` events with complimentary filter via fusion-pose-sensor.js.
+ */
+export default class PoseSensor {
+  constructor(config) {
+    this.config = config;
+    this.sensor = null;
+    this.fusionSensor = null;
+    this._out = new Float32Array(4);
+
+    // Can be 'sensor' (using RelativeOrientationSensor) or
+    // 'devicemotion' (using devicemotion events via FusionPoseSensor),
+    // or `null` if not yet set.
+    this.api = null;
+
+    // Store any errors from Sensors for debugging purposes
+    this.errors = [];
+
+    // Quaternions for caching transforms
+    this._sensorQ = new Quaternion();
+    this._worldToScreenQ = new Quaternion();
+    this._outQ = new Quaternion();
+
+    this._onSensorRead = this._onSensorRead.bind(this);
+    this._onSensorError = this._onSensorError.bind(this);
+    this._onOrientationChange = this._onOrientationChange.bind(this);
+
+    this._onOrientationChange();
+    this.init();
+  }
+
+  init() {
+    // Attempt to use the RelativeOrientationSensor from Generic Sensor APIs.
+    // First available in Chrome M63, this can fail for several reasons, and attempt
+    // to fallback to devicemotion. Failure scenarios include:
+    //
+    // * Generic Sensor APIs do not exist; fallback to devicemotion.
+    // * Underlying sensor does not exist; no fallback possible.
+    // * Feature Policy failure (in an iframe); no fallback.
+    //   https://github.com/immersive-web/webxr/issues/86
+    // * Permission to sensor data denied; respect user agent; no fallback to devicemotion.
+    //   Browsers are heading towards disabling devicemotion when sensors are denied as well.
+    //   https://www.chromestatus.com/feature/5023919287304192
+    let sensor = null;
+    try {
+      sensor = new RelativeOrientationSensor({ frequency: SENSOR_FREQUENCY });
+      sensor.addEventListener('error', this._onSensorError);
+    } catch (error) {
+      this.errors.push(error);
+
+      // Sensors are available in Chrome M63, however the Feature Policy
+      // integration is not available until Chrome M65, resulting in Sensors
+      // only being available in main frames.
+      // https://developers.google.com/web/updates/2017/09/sensors-for-the-web#feature_policy_integration
+      if (error.name === 'SecurityError') {
```

Comment:
@cvan just discovered that Sensors are only top level until Chrome M65 -- what do you think about enabling devicemotion in this case, at least for now? Otherwise this would break users on Chrome 63+, or Chrome 65+ without appropriate permissions.

Reply:
To try this out, I added permissions to the iframe example, which requires the devicemotion fallback on Chrome 63

Reply:
ah, good catch here